build-tools/automation/build.linux.groovy

```diff
@@ -0,0 +1,220 @@
+                        waitSecondsBeforeRetry = waitSecondsBeforeRetry * 2
+                    }
+
+                    retryAttempt++
+                    scmVars = checkout scm
+                }
+            }
+        }
+
+        def workspace = "/mnt/jenkins/workspace/${env.JOB_BASE_NAME}"
+        utils = load "${workspace}/${XADir}/build-tools/automation/utils.groovy"
+
+        utils.stageWithTimeout('init', 30, 'MINUTES', XADir, true) {    // Typically takes less than a second for CI builds; execChRootCommand execution can take several minutes for PR builds
+            // Note: PR plugin environment variable settings available here: https://wiki.jenkins.io/display/JENKINS/GitHub+pull+request+builder+plugin
+            isPr = env.ghprbActualCommit != null
+            isStable = env.IsStable == '1'
+            def branch = isPr ? env.GIT_BRANCH : scmVars.GIT_BRANCH
+            def commit = isPr ? env.ghprbActualCommit : scmVars.GIT_COMMIT
+
+            def buildType = isPr ? 'PR' : 'CI'
+
+            echo "HostName: ${env.NODE_NAME}"
+            echo "Git repo: ${env.GitRepo}"     // Defined as an environment variable in the jenkins build definition
+            echo "Job: ${env.JOB_BASE_NAME}"
+            echo "Branch: ${branch}"
+            echo "Commit: ${commit}"
+            echo "Build type: ${buildType}"
+            echo "Stable build workflow: ${isStable}"
+
+            pBuilderBindMounts = "/home/${env.USER}"
+            echo "pBuilderBindMounts: ${pBuilderBindMounts}"
+
+            if (isPr) {
+                echo "PR id: ${env.ghprbPullId}"
+                echo "PR link: ${env.ghprbPullLink}"
+
+                // Clear out the PR title and description. This is the equivalent of $JENKINS_HOME/global-pre-script/remove-problematic-ghprb-parameters.groovy used by freestyle builds
+                echo "Clearing the PR title and description environment variables to avoid any special characters contained within from tripping up the build"
+                env.ghprbPullTitle = ''
+                env.ghprbPullLongDescription = ''
+
+                chRootPackages = 'mono-devel autoconf automake build-essential vim-common p7zip-full cmake gettext libtool libgdk-pixbuf2.0-dev intltool pkg-config ruby scons wget xz-utils git nuget ca-certificates-mono g++-mingw-w64 gcc-mingw-w64 libzip4 openjdk-8-jdk unzip lib32stdc++6 lib32z1 libtinfo-dev:i386 linux-libc-dev:i386 zlib1g-dev:i386 gcc-multilib g++-multilib referenceassemblies-pcl zip libz-mingw-w64-dev msbuild mono-csharp-shell libsqlite3-dev sqlite3 cli-common-dev curl'
+
+                if (utils.hasPrLabel(env.GitRepo, env.ghprbPullId, 'full-mono-integration-build')) {
+                    hasPrLabelFullMonoIntegrationBuild = true
+                    buildTarget = 'jenkins'
+                } else {
+                    buildTarget = 'all'
+                }
+
+                execChRootCommand(env.ChRootName, chRootPackages, pBuilderBindMounts,
```

Comment:
You should generally only have one invocation of `chroot` in the whole job as it'll throw away the environment between each chroot invocation and creating a new chroot is pretty expensive.

I'd wrap the whole script into `chroot` and do the `stageWithTimeout` inside of it, that also matches how it's done in the freestyle builds.

Reply:
As we discussed, this is a bit tricky. The chroot command only supports executing a command (or set of commands) via a text string. In other words, it doesn't serve as a block command that can contain other groovy code as part of a body.
Since the primary purpose for the pipeline build is to provide an indication when the build has completed and packages published prior to testing, the pipeline steps have been consolidated between a build and test stage.
commit https://github.com/xamarin/xamarin-android/pull/2838/commits/1cfe44bb4dac27c9db88379700eeaa1f1fa97880